File: data_generator/generator.py
#!/usr/bin/python
import csv
import sys
from scipy.stats import norm
from scipy.stats import poisson
import numpy as np
import os

def read_specification(spec_file):
    with open(spec_file, 'rb') as csv_file:
        reader = csv.DictReader(csv_file, delimiter=',')
        spec = reader.next()  # expecting single row
        return spec


def generate_data(type, number, mean, sd):
    if type == 'gaussian':
        return norm.rvs(loc=mean, scale=sd, size=number)
    elif type == 'poisson':
        return poisson.rvs(mean, size=number)
    else:
        raise("wrong distribution type", type)


def write(with_intervals, data, out_filename):
    if with_intervals:
        fieldnames=['value', 'interval']
    else:
        fieldnames=['value']
    with open(out_filename, 'wb') as out_file:
        csv_writer = csv.DictWriter(out_file, delimiter='\t', fieldnames=fieldnames)
        csv_writer.writeheader()
        for row in data:
            if with_intervals:
                csv_writer.writerow(dict(value=row, interval=1))
            else:
                csv_writer.writerow(dict(value=row))


def generate_dataset(spec, out_dir):
    description = spec['Description']
    should_add_intervals = spec['NormalType'] == "poisson" or spec['AnomaliesType'] == "poisson"
    normal_type = spec['NormalType']
    normal_count = int(spec['NormalCount'])
    normal_mean = float(spec['NormalMean'])
    normal_sd = float(spec['NormalSd'])
    anomalies_type = spec['AnomaliesType']
    anomalies_count = int(spec['AnomaliesCount'])
    anomalies_mean = float(spec['AnomaliesMean'])
    anomalies_sd = float(spec['AnomailesSd'])
    normal_data = generate_data(normal_type, normal_count, normal_mean, normal_sd)
    anomalies_data = generate_data(anomalies_type, anomalies_count, anomalies_mean, anomalies_sd)
    dest_dir = os.path.join(out_dir, description)
    if not os.path.exists(dest_dir):
        os.makedirs(dest_dir)
    write(should_add_intervals, normal_data, os.path.join(dest_dir, 'normal'))
    write(should_add_intervals, anomalies_data, os.path.join(dest_dir, 'anomalous'))
    write(should_add_intervals, np.append(normal_data, anomalies_data), os.path.join(dest_dir, 'merged'))


if __name__ == '__main__':
    if len(sys.argv) != 3:
        print "Usage:", sys.argv[0], "<spec_file> <out_dir>"
        exit(1)
    spec_file = sys.argv[1]
    out_dir = sys.argv[2]
    generate_dataset(read_specification(spec_file), out_dir)

File: data_generator/generator.R
#!/usr/bin/Rscript

#
# Called as:
#   generator.R <spec_file> <output_dir>
#

#Description,NormalType,AnomaliesType,NormalCount,AnomaliesCount,NormalMean,AnomaliesMean,NormalSd,AnomailesSd
description <- function(x) { x[1] }
normal_type <- function(x) { x[2] }
anomalies_type<- function(x) { x[3] }
normal_count <- function(x) { x[4] }
anomalies_count <- function(x) { x[5] }
normal_mean <- function(x) { x[6] }
anomalies_mean <- function(x) { x[7] }
normal_sd <- function(x) { x[8] }
anomalies_sd <- function(x) { x[9] }

read_specification <- function(filename) {
    read.table(filename, sep=",", header=TRUE)
}

generate_data <- function(type, number, mean, sd) {
    if (type == 'gaussian') {
        values <- rnorm(number, mean=mean, sd=sd)
        data.frame(value=values)
    } else if (type == 'poisson') {
        values <- rpois(number, lambda=mean)
        data.frame(value=values)
    } else {
        stop("wrong distribution type")
    }
}

maybe_add_intervals <- function(should_add, data) {
    if (should_add) {
        rows <- nrow(data)
        intervals <- rep(1, rows)
        new_values <- sapply(data$value, as.integer)
        data.frame(value=new_values, interval=intervals)
    } else {
        data
    }
}

generate_dataset <- function(x, out_dir="generated/") {
    should_add_intervals <-
        normal_type(x)    == "poisson" ||
        anomalies_type(x) == "poisson"
    normal_data <-
        maybe_add_intervals(should_add_intervals,
                            generate_data(normal_type(x),
                                          as.numeric(normal_count(x)),
                                          as.numeric(normal_mean(x)),
                                          as.numeric(normal_sd(x))))
    normal_data2 <-
        maybe_add_intervals(should_add_intervals,
                            generate_data(normal_type(x),
                                          as.numeric(normal_count(x)),
                                          as.numeric(normal_mean(x)),
                                          as.numeric(normal_sd(x))))

    anomalous_data <-
        maybe_add_intervals(should_add_intervals,
                            generate_data(anomalies_type(x),
                                          as.numeric(anomalies_count(x)),
                                          as.numeric(anomalies_mean(x)),
                                          as.numeric(anomalies_sd(x))))
    if (! file.exists(out_dir)) {
        dir.create(out_dir)
    }
    dirname <- file.path(out_dir, description(x))
    if (! file.exists(dirname)) {
        dir.create(dirname)
    }
    #message(paste("Writing", description(x), "into", dirname))
    write.table(normal_data, paste(dirname, "/normal", sep=""), row.names=FALSE, sep="\t", quote=FALSE)
    write.table(normal_data2, paste(dirname, "/normal2", sep=""), row.names=FALSE, sep="\t", quote=FALSE)
    write.table(anomalous_data, paste(dirname, "/anomalous", sep=""), row.names=FALSE, sep="\t", quote=FALSE)
    write.table(rbind(normal_data, anomalous_data), paste(dirname, "/merged", sep=""), row.names=FALSE, sep="\t", quote=FALSE)
}

main <- function() {
    args <- commandArgs(trailingOnly = TRUE)
    spec_file <- args[1]
    out_dir   <- args[2]
    whatever  <- apply(read_specification(spec_file), 1, function(x) { generate_dataset(x, out_dir=out_dir) })
}

main()